File: TOURS_MS/model/Keypoint.go
package model

import "encoding/json"

type Keypoint struct {
	Id          int     `bson:"_id,omitempty" json:"id"`
	TourId      int     `bson:"tour_id,omitempty" json:"tour_id"`
	Name        string  `bson:"name,omitempty" json:"name"`
	Latitude    float64 `bson:"latitude,omitempty" json:"latitude"`
	Longitude   float64 `bson:"longitude,omitempty" json:"longitude"`
	Description string  `bson:"description,omitempty" json:"description"`
	Position    int     `bson:"position,omitempty" json:"position"`
	Image       string  `bson:"image,omitempty" json:"image"`
	Secret      string  `bson:"secret,omitempty" json:"secret"`
}

func (k *Keypoint) UnmarshalJSON(data []byte) error {

	var temp struct {
		Id          int
		TourId      int
		Name        string
		Latitude    float64
		Longitude   float64
		Description string
		Position    int
		Image       string
		Secret      string
	}

	if err := json.Unmarshal(data, &temp); err != nil {
		return err
	}

	k.Id = temp.Id
	k.TourId = temp.TourId
	k.Name = temp.Name
	k.Latitude = temp.Latitude
	k.Longitude = temp.Longitude
	k.Description = temp.Description
	k.Position = temp.Position
	k.Image = temp.Image
	k.Secret = temp.Secret

	return nil
}

func (k *Keypoint) MarshalJSON() ([]byte, error) {
	return json.Marshal(struct {
		Id          int
		TourId      int
		Name        string
		Latitude    float64
		Longitude   float64
		Description string
		Position    int
		Image       string
		Secret      string
	}{
		Id:          k.Id,
		TourId:      k.TourId,
		Name:        k.Name,
		Latitude:    k.Latitude,
		Longitude:   k.Longitude,
		Description: k.Description,
		Position:    k.Position,
		Image:       k.Image,
		Secret:      k.Secret,
	})
}

File: BACK/src/Modules/Tours/Explorer.Tours.Core/Domain/Keypoint.cs
using Explorer.BuildingBlocks.Core.Domain;

namespace Explorer.Tours.Core.Domain;

public class Keypoint : Entity
{
    public Keypoint()
    {
    }

    public Keypoint(long tourId, string name, double latitude, double longitude, string? description, int? position,
        string? image, string? secret)
    {
        Validate(name, latitude, longitude, position);

        TourId = tourId;
        Name = name;
        Latitude = latitude;
        Longitude = longitude;
        Description = description;
        Position = position;
        Image = image;
        Secret = secret;
    }

    public long TourId { get; }
    public Tour Tour { get; }
    public string Name { get; init; }
    public double Latitude { get; init; }
    public double Longitude { get; init; }
    public string? Description { get; init; }
    public int? Position { get; init; }
    public string? Image { get; set; }
    public string? Secret { get; set; }

    private static void Validate(string name, double latitude, double longitude, int? position)
    {
        if (string.IsNullOrWhiteSpace(name)) throw new ArgumentException("Invalid name");
        if (latitude is > 90 or < -90) throw new ArgumentException("Invalid latitude");
        if (longitude is > 180 or < -180) throw new ArgumentException("Invalid longitude");
        if (position < 1) throw new ArgumentException("Invalid position");
    }
}